kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceAction.java

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.kafkarest.resources.v3;
+
+import static java.util.Objects.requireNonNull;
+
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.protobuf.ByteString;
+import io.confluent.kafka.serializers.subject.strategy.SubjectNameStrategy;
+import io.confluent.kafkarest.controllers.ProduceController;
+import io.confluent.kafkarest.controllers.RecordSerializer;
+import io.confluent.kafkarest.controllers.SchemaManager;
+import io.confluent.kafkarest.entities.EmbeddedFormat;
+import io.confluent.kafkarest.entities.ProduceResult;
+import io.confluent.kafkarest.entities.RegisteredSchema;
+import io.confluent.kafkarest.entities.v3.ProduceRequest;
+import io.confluent.kafkarest.entities.v3.ProduceRequest.ProduceRequestData;
+import io.confluent.kafkarest.entities.v3.ProduceRequest.ProduceRequestHeader;
+import io.confluent.kafkarest.entities.v3.ProduceResponse;
+import io.confluent.kafkarest.entities.v3.ProduceResponse.ProduceResponseData;
+import io.confluent.kafkarest.exceptions.StatusCodeException;
+import io.confluent.kafkarest.response.StreamingResponse;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collector;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response.Status;
+
+@Path("/v3/clusters/{clusterId}/topics/{topicName}/records")
```

Comment:
All V3 APIs use the same prefix `/v3/clusters/{clusterId}`, so we are tied to it. It would be highly inconsistent to diverge here. The `/topics/{topicName}` seems indispensable too, since we need to know the topic name. Up for discussion really is what comes after it, `/records`. I used only `/records` exactly to make the API shorter. But, if we were to be super consistent with everything else we have done for V3, this API should really be:

`POST /v3/clusters/{clusterId}/topics/{topicName}/partitions/-/records`

Why? Because I also want to add:

* `POST /v3/clusters/{clusterId}/topics/{topicName}/partitions/{partitionId}/records` to produce to a single partition (i.e. no partition_id field in the request body).
* `GET /v3/clusters/{clusterId}/topics/{topicName}/partitions/{partitionId}/records` to list records in a partition.
* `GET /v3/clusters/{clusterId}/topics/{topicName}/partitions/{partitionId}/records/{offset}` to get a single record from a partition.